[Mac] Add a pop-up menu for selecting caption style profiles

jernoble opened the pull request:
#### 6711229d438fbd977d5b1b9b572096382597e08f
<pre>
[Mac] Add a pop-up menu for selecting caption style profiles
<a href="https://rdar.apple.com/163067314">rdar://163067314</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=301134">https://bugs.webkit.org/show_bug.cgi?id=301134</a>

Reviewed by Eric Carlson.

Add a default implementation of a pop-up menu that allows the user to choose
between their current caption style profiles. The menu will source profile
names through MediaAccessibility soft-linked methods, allowing the entire
backing for this menu to be stubbed and tested by TestWebKit API.

Test: Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::platformProfileIDs):
(WebCore::CaptionUserPreferencesMediaAF::platformActiveProfileID):
(WebCore::CaptionUserPreferencesMediaAF::canSetActiveProfileID):
(WebCore::CaptionUserPreferencesMediaAF::setActiveProfileID):
(WebCore::CaptionUserPreferencesMediaAF::nameForProfileID):
* Source/WebCore/page/CaptionUserPreferencesMediaAF.h:
* Source/WebCore/platform/cf/MediaAccessibilitySoftLink.cpp:
* Source/WebCore/platform/cf/MediaAccessibilitySoftLink.h:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuController.h: Copied from Tools/TestWebKitAPI/SoftLinkShim.h.
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuController.mm: Added.
(-[WKCaptionStyleMenuController init]):
(-[WKCaptionStyleMenuController rebuildMenu]):
(-[WKCaptionStyleMenuController captionStyleMenu]):
(-[WKCaptionStyleMenuController profileMenuItemSelected:]):
(-[WKCaptionStyleMenuController profileMenuItemHighlighted:]):
(-[WKCaptionStyleMenuController systemCaptionStyleSettingsItemSelected:]):
(-[WKCaptionStyleMenuController menuWillOpen:]):
(-[WKCaptionStyleMenuController menu:willHighlightItem:]):
(-[WKCaptionStyleMenuController menuDidClose:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SoftLinkShim.h:
(TestWebKitAPI::SoftLinkShim::SoftLinkShim):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm:
(TestWebKitAPI::MediaAccessibilityShim::shimmedMACaptionAppearanceSetActiveProfileID):
(TestWebKitAPI::MediaAccessibilityShim::shimmedMACaptionAppearanceCopyProfileName):
(TestWebKitAPI::MediaAccessibilityShim::resetToDefaultValues):
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenu)):
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenuHighlight)):
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenuSelect)):

Canonical link: <a href="https://commits.webkit.org/301952@main">https://commits.webkit.org/301952@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/671407a0e5a57bd9139fa94a76021ffef7209c4d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127442 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47090 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38228 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134518 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78997 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2b6bbd28-5361-4c2a-8484-6e7153339872) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129314 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47706 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55613 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97014 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64965 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/95e21234-aac3-43f9-b23c-0f0d73d6fccf) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130390 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38168 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114143 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77494 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d53d349a-1958-45f2-9285-b4619499ab2b) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37001 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32248 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77892 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108027 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32668 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137003 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54101 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41689 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105534 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54612 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110492 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105187 "Found 1 new API test failure: WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26847 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50700 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29157 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51680 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54038 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60125 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53272 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56729 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55031 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
```diff
@@ -1008,7 +1008,48 @@ Vector<RefPtr<TextTrack>> CaptionUserPreferencesMediaAF::sortedTrackListForMenu(
-    
+
+Vector<String> CaptionUserPreferencesMediaAF::platformProfileIDs()
+    {
```
Nit: odd indentation

eric-carlson commented on Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuController.mm:
```diff
@@ -0,0 +1,148 @@
+
+- (void)menu:(NSMenu *)menu willHighlightItem:(NSMenuItem *)item
+{
+    if ([item.representedObject isKindOfClass:NSString.class])
```
Do you need to check the object class both here and in `profileMenuItemHighlighted`?

eric-carlson commented on Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuControllerMac.mm:
```diff
@@ -0,0 +1,148 @@
+
+- (void)menuDidClose:(NSMenu *)menu
+{
+    if (!_savedActiveProfileID.isEmpty())
+        CaptionUserPreferencesMediaAF::setActiveProfileID(_savedActiveProfileID);
```
It might be worth only setting the active profile when `_savedActiveProfileID` changes to avoid spamming the system with a changed notification

jernoble committed:
[Mac] Add a pop-up menu for selecting caption style profiles
rdar://163067314
https://bugs.webkit.org/show_bug.cgi?id=301134

Reviewed by Eric Carlson.

Add a default implementation of a pop-up menu that allows the user to choose
between their current caption style profiles. The menu will source profile
names through MediaAccessibility soft-linked methods, allowing the entire
backing for this menu to be stubbed and tested by TestWebKit API.

Test: Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::platformProfileIDs):
(WebCore::CaptionUserPreferencesMediaAF::platformActiveProfileID):
(WebCore::CaptionUserPreferencesMediaAF::canSetActiveProfileID):
(WebCore::CaptionUserPreferencesMediaAF::setActiveProfileID):
(WebCore::CaptionUserPreferencesMediaAF::nameForProfileID):
* Source/WebCore/page/CaptionUserPreferencesMediaAF.h:
* Source/WebCore/platform/cf/MediaAccessibilitySoftLink.cpp:
* Source/WebCore/platform/cf/MediaAccessibilitySoftLink.h:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuController.h: Copied from Tools/TestWebKitAPI/SoftLinkShim.h.
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuController.mm: Added.
(-[WKCaptionStyleMenuController init]):
(-[WKCaptionStyleMenuController rebuildMenu]):
(-[WKCaptionStyleMenuController captionStyleMenu]):
(-[WKCaptionStyleMenuController profileMenuItemSelected:]):
(-[WKCaptionStyleMenuController profileMenuItemHighlighted:]):
(-[WKCaptionStyleMenuController systemCaptionStyleSettingsItemSelected:]):
(-[WKCaptionStyleMenuController menuWillOpen:]):
(-[WKCaptionStyleMenuController menu:willHighlightItem:]):
(-[WKCaptionStyleMenuController menuDidClose:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SoftLinkShim.h:
(TestWebKitAPI::SoftLinkShim::SoftLinkShim):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm:
(TestWebKitAPI::MediaAccessibilityShim::shimmedMACaptionAppearanceSetActiveProfileID):
(TestWebKitAPI::MediaAccessibilityShim::shimmedMACaptionAppearanceCopyProfileName):
(TestWebKitAPI::MediaAccessibilityShim::resetToDefaultValues):
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenu)):
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenuHighlight)):
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenuSelect)):

Canonical link: https://commits.webkit.org/301952@main
